numpy/core/src/common/npy_cpu.c

```diff
@@ -0,0 +1,219 @@
+                                            cpu_features[NPY_OPT_AVX512VL];
+    /* Cannon Lake */
+    cpu_features[NPY_OPT_AVX512IFMA]      = (reg[1] & (1 << 21)) != 0;
+    cpu_features[NPY_OPT_AVX512VBMI]      = (reg[2] & (1 << 1))  != 0;
+    cpu_features[NPY_OPT_XCNL]            = cpu_features[NPY_OPT_XSKX] &&
+                                            cpu_features[NPY_OPT_AVX512IFMA] &&
+                                            cpu_features[NPY_OPT_AVX512VBMI];
+    /* Cascade Lake */
+    cpu_features[NPY_OPT_AVX512VNNI]      = (reg[2] & (1 << 11)) != 0;
+    cpu_features[NPY_OPT_XCEL]            = cpu_features[NPY_OPT_XCNL] &&
+                                            cpu_features[NPY_OPT_AVX512VNNI];
+
+    /* Ice Lake */
+    cpu_features[NPY_OPT_AVX512VBMI2]     = (reg[2] & (1 << 6))  != 0;
+    cpu_features[NPY_OPT_AVX512BITALG]    = (reg[2] & (1 << 12)) != 0;
+    cpu_features[NPY_OPT_XICL]            = cpu_features[NPY_OPT_XCEL] &&
+                                            cpu_features[NPY_OPT_AVX512VBMI2] &&
+                                            cpu_features[NPY_OPT_AVX512BITALG] &&
+                                            cpu_features[NPY_OPT_AVX512VPOPCNTDQ];
+}
+/***************** POWER ******************/
+#elif defined(NPY_CPU_PPC64) || defined(NPY_CPU_PPC64LE)
+
+#if defined __linux__
+    #include "sys/auxv.h"
+    #ifndef AT_HWCAP2
+        #define AT_HWCAP2 26
+    #endif
+    #ifndef PPC_FEATURE2_ARCH_3_00
+        #define PPC_FEATURE2_ARCH_3_00 0x00800000
+    #endif
+#endif
+
+void npy_cpu_init(void)
+{
+#if defined __linux__
+    unsigned long long hwcap;
+#endif
+    memset(cpu_features, 0, sizeof(cpu_features[0]) * NPY_OPT_MAX);
+
+#if defined __linux__
+    hwcap = getauxval(AT_HWCAP);
+    if ((hwcap & PPC_FEATURE_HAS_VSX) == 0)
+        return;
+
+    hwcap = getauxval(AT_HWCAP2);
+    if (hwcap & PPC_FEATURE2_ARCH_3_00)
+    {
+        cpu_features[NPY_OPT_VSX]  =
+        cpu_features[NPY_OPT_VSX2] =
+        cpu_features[NPY_OPT_VSX3] = 1;
+        return;
+    }
+    cpu_features[NPY_OPT_VSX2] = (hwcap & PPC_FEATURE2_ARCH_2_07) != 0;
+    cpu_features[NPY_OPT_VSX]  = 1;
+
+/* TODO: Check VSX availability in runtime for other platforms */
+#else
+    #ifdef NPY_CPU_PPC64LE
+    cpu_features[NPY_OPT_VSX]  = 1;
+    cpu_features[NPY_OPT_VSX2] = 1;
+    #else
+    cpu_features[NPY_OPT_VSX]  = 1;
+    #endif
+#endif
+}
+
+/***************** ARM ******************/
+/* todo */
+#elif defined(__arm__) || defined(__aarch64__)
+#include <stdio.h>
+void npy_cpu_init(void)
+{
+    memset(cpu_features, 0, sizeof(cpu_features[0]) * NPY_OPT_MAX);
+    fprintf(stderr, "NumPy: ARM runtime check not implemented yet\n");
+}
+
+/*********** Unsupported ARCH ***********/
+
+#else
+#include <stdio.h>
+void npy_cpu_init(void)
+{
+    memset(cpu_features, 0, sizeof(cpu_features[0]) * NPY_OPT_MAX);
+    fprintf(stderr, "NumPy: Unsupported architecture\n");
```

Comment:
If this is runtime, raise a python error, results needs to be checked in call. Why not just set the python error and return -1 on error, 0 otherwise and handle error on return?

Reply:
Also above. No C print statements except for debug code.

Reply:
I removed it instead, set all features to zero is more than enough but we will need to add python level functions to check availability of CPU features and also to disable/enable specific feature on runtime.